OpenRA.Game/Graphics/HardwareCursor.cs

```diff
@@ -38,74 +37,98 @@ public HardwareCursor(CursorProvider cursorProvider)
 				hardwarePalette.AddPalette(p.Key, p.Value, false);
 
 			hardwarePalette.Initialize();
-
 			sheetBuilder = new SheetBuilder(SheetType.Indexed);
 			foreach (var kv in cursorProvider.Cursors)
 			{
 				var palette = cursorProvider.Palettes[kv.Value.Palette];
-				var hc = kv.Value.Frames
-					.Select(f => CreateCursor(f, palette, kv.Key, kv.Value))
-					.ToArray();
+				var frames = kv.Value.Frames;
+				var frameData = new byte[frames.Length][];
+				var frameHotspots = new int2[frames.Length];
+				var cursors = new IHardwareCursor[frames.Length];
+				var frameSprites = new Sprite[frames.Length];
+
+				// Hardware cursors have a number of odd platform-specific bugs/limitations.
+				// Reduce the number of edge cases by padding the individual frames such that:
+				// - the hotspot is inside the frame bounds (enforced by SDL)
+				// - all frames within a sequence have the same size (needed for macOS 10.15)
+				// - the frame size is a multiple of 8 (needed for Windows)
+				Rectangle sequenceBounds = Rectangle.FromLTRB(0, 0, 1, 1);
+				for (var i = 0; i < frames.Length; i++)
+				{
+					var frame = frames[i];
+
+					frameSprites[i] = sheetBuilder.Add(frame.Data, frame.Size, 0, frame.Offset);
+
+					// CreateHardwareCursor expects data in RGBA
+					frameData[i] = FrameToRGBA(frame, palette);
+
+					// Calculate frame bounds relative to the hotspot position
+					frameHotspots[i] = kv.Value.Hotspot - frame.Offset.ToInt2() + new int2(frame.Size) / 2;
+
+					sequenceBounds = Rectangle.Union(sequenceBounds, new Rectangle(-frameHotspots[i], frame.Size));
+				}
 
-				hardwareCursors.Add(kv.Key, hc);
+				var paddedSize = 8 * new int2((sequenceBounds.Width + 7) / 8, (sequenceBounds.Height + 7) / 8);
+				for (var i = 0; i < frames.Length; i++)
```

Comment:
Im guessing that we need to do this in two stages because how `paddedSize` is calculated, right?

Reply:
Correct. I'll move the `frameSprites` and `frameData` handling into this loop to make the separation of concerns clearer.